File: _fan_probe.py
# a2.83 investigation. applyShading opens with a full geo.computeVertexNormals()
# whose answer is then overwritten by the per-vertex union-find - except where
# the union-find CONTINUEs. _prof_probe says that baseline is 12-15% of a warm
# shade, so the question worth answering before touching anything is: WHO still
# reads it?
#
# This instruments a COPY of index.html (index.html is never touched) to count,
# for every shade, exactly which attribute vertices the write loop leaves alone
# and what the baseline is giving them.
import io, os, shutil, tempfile, sys, subprocess, threading
import http.server, socketserver

ROOT = r'C:\Users\a.bodrov\Projects\kubik'
# A FRESH profile of this probe's own. Sharing Chrome's default profile
# makes the run exit in seconds whenever a Chrome is open; a FIXED
# directory makes the next run open this run's autosaved scene.
PROF = tempfile.mkdtemp(prefix='_prof_')
CHROME = r'C:\Program Files\Google\Chrome\Application\chrome.exe'
PORT = 8862
OUT = '_fan_out.txt'

src = io.open(os.path.join(ROOT, 'index.html'), encoding='utf-8').read()

HEAD_ANCHOR = "    const normals = geo.attributes.normal.array;"
HEAD_ADD = """
    /* ---- INJECTED BY _fan_probe.py - not in the shipped file ---- */
    var _F = window.__FAN = window.__FAN || { shades: 0, verts: 0, used: 0,
      noGroup: 0, notEnrolled: 0, noAcc: 0, degen: 0, unwritten: 0,
      degenBaseZero: 0, degenBaseNonZero: 0, degenIslands: 0,
      unwrittenBaseZero: 0, unwrittenBaseNonZero: 0, maxBaseGap: 0 };
    var _base = normals.slice();
    var _wrote = new Uint8Array(posAttr.count);
"""

LOOP_OLD = """      const ais = logicalGroups[l];
      for (let i = 0; i < ais.length; i++) {
        const ai = ais[i], g = groupOfAttr[ai];
        if (g < 0 || ufStamp[g] !== stamp) continue;
        const r = find(g);
        if (accStamp[r] !== stamp) continue;
        const x = accX[r], y = accY[r], z = accZ[r];
        if (x * x + y * y + z * z < 1e-12) continue;   // degenerate fan: keep the baseline
        normals[ai * 3] = x; normals[ai * 3 + 1] = y; normals[ai * 3 + 2] = z;
      }"""

LOOP_NEW = """      const ais = logicalGroups[l];
      for (let i = 0; i < ais.length; i++) {
        const ai = ais[i], g = groupOfAttr[ai];
        if (g < 0) { _F.noGroup++; continue; }
        if (ufStamp[g] !== stamp) { _F.notEnrolled++; continue; }
        const r = find(g);
        if (accStamp[r] !== stamp) { _F.noAcc++; continue; }
        const x = accX[r], y = accY[r], z = accZ[r];
        if (x * x + y * y + z * z < 1e-12) {
          _F.degen++;
          _F.degenIslands += touched.length;
          var _bx = _base[ai * 3], _by = _base[ai * 3 + 1], _bz = _base[ai * 3 + 2];
          if (_bx * _bx + _by * _by + _bz * _bz < 1e-12) _F.degenBaseZero++;
          else _F.degenBaseNonZero++;
          continue;
        }
        _wrote[ai] = 1;
        /* HOW FAR THE BASELINE WAS FROM THE ANSWER. If these agreed the
           baseline would be harmless as a fallback; where they disagree it
           is a second shading model answering for the first. */
        var _dx = _base[ai * 3] - x, _dy = _base[ai * 3 + 1] - y, _dz = _base[ai * 3 + 2] - z;
        var _gap = Math.sqrt(_dx * _dx + _dy * _dy + _dz * _dz);
        if (_gap > _F.maxBaseGap) _F.maxBaseGap = _gap;
        normals[ai * 3] = x; normals[ai * 3 + 1] = y; normals[ai * 3 + 2] = z;
      }"""

TAIL_ANCHOR = "    geo.attributes.normal.needsUpdate = true;"
TAIL_ADD = """    {
      /* WHO IS LEFT. Only vertices the INDEX actually references matter -
         an unused attribute vertex is never rasterised. */
      _F.shades++;
      _F.verts += posAttr.count;
      var _used = new Uint8Array(posAttr.count);
      for (var _u = 0; _u < index.count; _u++) _used[index.getX(_u)] = 1;
      for (var _v = 0; _v < posAttr.count; _v++) {
        if (!_used[_v]) continue;
        _F.used++;
        if (_wrote[_v]) continue;
        _F.unwritten++;
        var _qx = _base[_v * 3], _qy = _base[_v * 3 + 1], _qz = _base[_v * 3 + 2];
        if (_qx * _qx + _qy * _qy + _qz * _qz < 1e-12) _F.unwrittenBaseZero++;
        else _F.unwrittenBaseNonZero++;
      }
    }
"""

ENTRY_ANCHOR = """  // Safe baseline: if anything below throws, the mesh is still shaded.
  geo.computeVertexNormals();"""
ENTRY_ADD = """  {
    /* ---- INJECTED BY _fan_probe.py ---- */
    var _G = window.__FAN0 = window.__FAN0 || { calls: 0, noAttr: 0, allZero: 0, wrongSize: 0, live: 0 };
    _G.calls++;
    var _na = geo.attributes.normal;
    if (!_na) _G.noAttr++;
    else if (geo.attributes.position && _na.count !== geo.attributes.position.count) _G.wrongSize++;
    else {
      var _z = true;
      for (var _i = 0; _i < _na.array.length; _i++) if (_na.array[_i] !== 0) { _z = false; break; }
      if (_z) _G.allZero++; else _G.live++;
    }
  }
"""

for name, old, new in (('entry', ENTRY_ANCHOR, ENTRY_ADD + ENTRY_ANCHOR),
                       ('head', HEAD_ANCHOR, HEAD_ANCHOR + HEAD_ADD),
                       ('loop', LOOP_OLD, LOOP_NEW),
                       ('tail', TAIL_ANCHOR, TAIL_ADD + TAIL_ANCHOR)):
    if src.count(old) != 1:
        print('%s ANCHOR MISS (%d)' % (name, src.count(old)))
        sys.exit(1)
    src = src.replace(old, new, 1)
print('instrumented the write loop')

JS = io.open(os.path.join(ROOT, '_fan_probe.js'), encoding='utf-8').read()
assert '</body>' in src
io.open(os.path.join(ROOT, '_fan_probe.html'), 'w', encoding='utf-8', newline='').write(
    src.replace('</body>', '<script>\n' + JS + '\n</script>\n</body>', 1))

result = {'text': None}
done = threading.Event()


class H(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *a, **kw):
        super().__init__(*a, directory=ROOT, **kw)

    def log_message(self, *a):
        pass

    def do_POST(self):
        n = int(self.headers.get('Content-Length', 0))
        result['text'] = self.rfile.read(n).decode('utf-8', 'replace')
        self.send_response(204)
        self.end_headers()
        done.set()


socketserver.TCPServer.allow_reuse_address = True
srv = socketserver.TCPServer(('127.0.0.1', PORT), H)
threading.Thread(target=srv.serve_forever, daemon=True).start()

url = 'http://127.0.0.1:%d/_fan_probe.html?debug=1' % PORT
cmd = [CHROME, '--headless=new', '--disable-gpu-sandbox', '--no-sandbox',
       '--enable-unsafe-swiftshader', '--use-angle=swiftshader',
       '--window-size=900,700', '--user-data-dir=' + PROF,
       # SHARED, and deliberately not the fresh profile: the app pulls three.js
       # from a CDN on every load, and a cold cache is what actually broke
       # _imp_probe. State fresh, network warm.
       '--disk-cache-dir=' + os.path.join(ROOT, '_httpcache'),
       '--no-first-run', '--no-default-browser-check', '--disable-sync',
       '--disable-background-networking', '--disable-component-update',
       '--disable-default-apps', '--disable-extensions', '--metrics-recording-only',
       '--mute-audio',
       url]
p = subprocess.Popen(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
ok = done.wait(timeout=180)
try:
    p.terminate(); p.wait(timeout=20)
except Exception:
    p.kill()
srv.shutdown()
shutil.rmtree(PROF, ignore_errors=True)

txt = result['text'] if ok and result['text'] else 'NO PROBE OUTPUT (timed out)'
io.open(os.path.join(ROOT, OUT), 'w', encoding='utf-8').write(txt)
print(txt)

File: _fan_probe.js
/* a2.83 investigation. Runs a battery of models through applyShading and
   reports, per model, which attribute vertices the union-find write loop
   leaves alone - because those, and only those, are what the opening
   geo.computeVertexNormals() is still paying for. */
(function () {
  var out = [], errs = [];
  var _ce = console.error;
  console.error = function () { errs.push(Array.prototype.join.call(arguments, ' ')); _ce.apply(console, arguments); };
  function log(k, v) { out.push(k + '=' + v); }
  var k, A, THREE;

  function snap() {
    var F = window.__FAN;
    return {
      used: F.used, unwritten: F.unwritten, degen: F.degen,
      noGroup: F.noGroup, notEnrolled: F.notEnrolled, noAcc: F.noAcc,
      dbz: F.degenBaseZero, dbn: F.degenBaseNonZero,
      ubz: F.unwrittenBaseZero, ubn: F.unwrittenBaseNonZero,
      gap: F.maxBaseGap
    };
  }
  function delta(a, b) {
    var d = {};
    for (var key in b) d[key] = key === 'gap' ? b[key] : b[key] - a[key];
    return d;
  }

  var rows = [];
  function measure(label, obj) {
    if (!obj) { rows.push(label + ': NOT BUILT'); return null; }
    var before = snap();
    k.applyShading(obj);
    var d = delta(before, snap());
    var tris = obj.mesh.geometry.index ? obj.mesh.geometry.index.count / 3 : 0;
    rows.push(label + ': ' + tris + 'tri ' + d.used + 'v' +
      ' | unwritten ' + d.unwritten +
      (d.unwritten ? ' (base zero ' + d.ubz + ', live ' + d.ubn + ')' : '') +
      ' | degen ' + d.degen +
      (d.degen ? ' (base zero ' + d.dbz + ', live ' + d.dbn + ')' : '') +
      (d.noGroup ? ' | noGroup ' + d.noGroup : '') +
      (d.notEnrolled ? ' | notEnrolled ' + d.notEnrolled : '') +
      (d.noAcc ? ' | noAcc ' + d.noAcc : ''));
    return d;
  }

  function fresh(name) {
    var o = k.createCubeObject(name, new THREE.Vector3(0, 0, 0));
    A.selectedObjectIds = new Set([o.id]);
    A.activeObjectId = o.id;
    k.refreshUI();
    return o;
  }
  function sub(o) {
    k.setMode('object');
    A.selectedObjectIds = new Set([o.id]);
    A.activeObjectId = o.id;
    A.selectedElements = new Set();
    k.subdivideSelection();
    if (!A.pendingOp) return false;
    k.flushPendingApply();
    k.confirmPendingOp();
    return true;
  }

  function main() {
    k = window.__kubik; A = k.App; THREE = k.THREE;

    log('0.instrumented', window.__FAN
      ? 'the write loop is counting; entry state ' + (window.__FAN0 ? 'too' : 'MISSING')
      : 'NO __FAN OBJECT - nothing was injected, every number below is meaningless');
    if (!window.__FAN) { return; }

    /* ---- 1. The battery. Every shape the app can make plus a couple it
       cannot, because the whole question is whether the baseline is ever
       READ, and one shape that reads it is enough to make dropping it a
       correctness change rather than a deletion. ---- */

    var cube = fresh('flat cube');
    measure('cube (flat)', cube);

    var sm = fresh('smooth cube');
    k.setMode('object');
    A.selectedObjectIds = new Set([sm.id]);
    A.activeObjectId = sm.id;
    sub(sm); sub(sm);
    measure('cube subdivided x2', sm);

    var kinds = [];
    try { kinds = Object.keys(k.PRIM_SPECS || {}); } catch (e) { kinds = []; }
    log('1.primitive_kinds', kinds.length ? kinds.join(', ') : 'PRIM_SPECS not reachable - primitives skipped');
    kinds.forEach(function (kind) {
      var spec = k.PRIM_SPECS[kind];
      var def = {};
      for (var f in (spec.def || {})) def[f] = spec.def[f];
      var o = null;
      try { o = k.createPrimitiveObject(kind, def, kind, new THREE.Vector3(0, 0, 0)); } catch (e) { }
      if (o) measure('prim ' + kind, o);
    });

    /* An OPEN surface: every boundary edge has one face, which the sharp
       rule forces sharp, so islands are as fragmented as they get. */
    var plane = k.createObjectFromEditable('plane', new THREE.Vector3(0, 0, 0), {
      positions: [0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1],
      groups: [{ triangles: [[0, 1, 2], [0, 2, 3]] }]
    }, k.makeMaterialSet(1), {});
    measure('open plane (one quad)', plane);

    /* THE FIXTURE THAT EXISTS TO TRIP IT. Two triangles in ONE face group,
       wound against each other, so at every one of their shared corners the
       island's summed triangle normal is exactly zero - which is the
       `degenerate fan` continue, the one branch that demonstrably reads the
       baseline. Union-find nodes are FACE GROUPS, so putting both triangles
       in one group is what guarantees they land in the same island; two
       groups would be split by the sharp test instead and prove nothing. */
    var fin = null;
    try {
      fin = k.createObjectFromEditable('fin', new THREE.Vector3(0, 0, 0), {
        positions: [0, 0, 0, 1, 0, 0, 0, 0, 1],
        groups: [{ triangles: [[0, 1, 2], [0, 2, 1]] }]
      }, k.makeMaterialSet(1), {});
    } catch (e) { errs.push('fin: ' + e); }
    var finD = measure('folded fin (one group, opposed winding)', fin);

    /* And the same fold with a real fan around it - a vertex where two
       opposed triangles meet AND four ordinary ones do, so the island sum
       cancels while the baseline, which sums every adjacent triangle
       regardless of island, does not. */
    var fan = null;
    try {
      fan = k.createObjectFromEditable('fan', new THREE.Vector3(0, 0, 0), {
        positions: [0, 0, 0, 1, 0, 0, 0, 0, 1, -1, 0, 0, 0, 0, -1, 0, 1, 0],
        groups: [
          { triangles: [[0, 1, 2], [0, 2, 1]] },
          { triangles: [[0, 3, 4], [0, 4, 5]] }
        ]
      }, k.makeMaterialSet(2), {});
    } catch (e) { errs.push('fan: ' + e); }
    measure('folded fin beside a live fan', fan);

    /* ---- 1b. AFTER THE OPS. Every op funnels through
       rebuildFromEditable, which makes a BRAND NEW BufferGeometry with no
       normal attribute - so an op is exactly the case where "whatever was
       in the array" is nothing at all, and any vertex the write loop skips
       would render black. ---- */
    function opCase(label, mode, sel, go) {
      var o = fresh('op ' + label);
      sub(o); sub(o);
      try {
        k.setMode(mode);
        A.selectedObjectIds = new Set([o.id]);
        A.activeObjectId = o.id;
        k.ensureHelpers(o);
        A.selectedElements = new Set(sel);
        go(o);
        if (A.pendingOp) { k.flushPendingApply(); k.confirmPendingOp(); }
      } catch (e) { rows.push('op ' + label + ': THREW ' + e); return; }
      measure('after ' + label, o);
    }
    opCase('extrude', 'face', [0], function () { k.extrudeSelection(); k.setPendingAmount(0.4); });
    opCase('inset', 'face', [0, 1], function () { k.insetSelection(); k.setPendingAmount(0.2); });
    opCase('bevel', 'edge', [0], function () { k.bevelSelection(); k.setPendingAmount(0.1); });
    opCase('subdivide', 'object', [], function () { k.subdivideSelection(); });
    opCase('solidify', 'object', [], function () { k.solidifySelection(); k.setPendingAmount(0.1); });
    opCase('array', 'object', [], function () { k.arraySelection(); k.setPendingAmount(1.5); });
    opCase('cleanup', 'object', [], function () { k.cleanupSelection(); });
    opCase('crease', 'edge', [0, 1, 2], function () { k.creaseSelection(); });
    opCase('shade smooth', 'object', [], function () { k.smoothByAngleSelection && k.smoothByAngleSelection(); });
    opCase('slide', 'edge', [4], function () { k.slideSelection(); k.setPendingAmount(0.3); });

    log('1.battery', rows.join('  ||  '));

    /* ---- 2. The verdict on unwritten vertices ---- */
    var F = window.__FAN;
    log('2.unwritten', F.unwritten === 0
      ? 'across every model above, the write loop wrote EVERY attribute vertex the index references - ' +
        F.used + ' of them - so nothing but the degenerate branch can read the baseline'
      : F.unwritten + ' of ' + F.used + ' referenced vertices were left to the baseline (' +
        F.unwrittenBaseZero + ' of them holding zero, ' + F.unwrittenBaseNonZero + ' holding a real normal)');
    log('2.why', 'noGroup ' + F.noGroup + ', notEnrolled ' + F.notEnrolled +
      ', noAcc ' + F.noAcc + ', degenerate fan ' + F.degen);

    /* ---- 3. Does the degenerate branch ever fire, and on what ---- */
    log('3.degenerate_fires', F.degen > 0
      ? 'yes - ' + F.degen + ' times, and the fixture built to cause it ' +
        (finD && finD.degen > 0 ? 'did' : 'DID NOT, so section 3 is about something else')
      : 'NEVER, on any model above - including the fixture built to cause it, which means the fixture is wrong, not the branch');
    log('3.what_the_baseline_gives', F.degen === 0 ? 'n/a'
      : F.degenBaseZero + ' of those kept a ZERO normal (the baseline cancelled too, so the fallback falls back to nothing) and ' +
        F.degenBaseNonZero + ' kept a live one summed across islands the shading model says are separate');
    log('3.islands_at_those_vertices', F.degen === 0 ? 'n/a'
      : (F.degenIslands / F.degen).toFixed(2) + ' islands on average at a degenerate vertex' +
        (F.degenIslands / F.degen > 1.01 ? ' - more than one, so the baseline is answering across a crease' : ''));

    /* ---- 4. Is the baseline even close to the answer where it IS
       overwritten? If it were, calling it a "safe baseline" would be fair.
       A unit-vector gap of 2 is a normal pointing the opposite way. ---- */
    log('4.max_baseline_gap', F.maxBaseGap.toFixed(4) +
      ' between two unit vectors, which is ' +
      (2 * Math.asin(Math.min(1, F.maxBaseGap / 2)) * 180 / Math.PI).toFixed(0) + ' degrees' +
      (F.maxBaseGap > 0.02
        ? ' - so the baseline is not an approximation of the shading answer, it is a different answer that happens to be overwritten'
        : ' - the two agree everywhere, which would make it a fair fallback'));

    /* ---- 5. What the normal attribute holds AT ENTRY. If the baseline is
       dropped this is what an unwritten vertex would keep. ---- */
    var G = window.__FAN0;
    log('5.entry_state', G
      ? G.calls + ' calls: no attribute at all ' + G.noAttr + ', wrong size ' + G.wrongSize +
        ', all zero ' + G.allZero + ', already holding normals ' + G.live
      : 'NOT INSTRUMENTED');
    log('5.means', !G ? 'n/a' : (G.noAttr > 0
      ? 'dropping the baseline needs an explicit create - ' + G.noAttr + ' calls arrived with no normal attribute'
      : 'the attribute always exists at entry, so dropping the baseline needs no allocation, only a value for every vertex'));

    log('console.errors', errs.length ? errs.join(' | ').slice(0, 400) : 'none');
  }

  var posted = false;
  function post() {
    if (posted) return;
    posted = true;
    try { fetch('/result', { method: 'POST', body: out.join('\n') }); }
    catch (e) { }
  }
  setTimeout(function () {
    if (!posted) { out.push('WATCHDOG=main did not finish - the last line above is where it hung'); post(); }
  }, 60000);
  function ready(cb, t) {
    t = t || 0;
    if (window.__kubik && window.__kubik.App) return cb();
    if (t > 400) { out.push('ERROR=no __kubik'); return post(); }
    setTimeout(function () { ready(cb, t + 1); }, 20);
  }
  window.addEventListener('error', function (e) { errs.push('onerror: ' + (e.message || e)); });
  setTimeout(function () {
    ready(function () {
      setTimeout(function () {
        try { main(); }
        catch (e) { out.push('ERROR=' + (e && e.stack ? e.stack.split('\n').slice(0, 8).join(' / ') : e)); }
        post();
      }, 800);
    });
  }, 400);
})();
